File: apps/ios/Voiid/Voiid/Main/LocationDetailView.swift
//
//  LocationDetailView.swift
//  Voiid
//
//  Full-screen detail for a location message (docs/LOCATION.md §4): the map, Open in Maps
//  (system handoff), and Directions. There is NO in-app routing and NO reverse-geocoding
//  to a street address (§10) — we hand off to the system map app, which keeps the
//  coordinate off any geocoder we control.
//
//  LIVE, not a snapshot. This view used to take a single `coordinate` and render it forever,
//  so a friend walking across town stayed pinned wherever they were when the sheet opened —
//  the fix stream was arriving the whole time and only the bubble showed it. It now observes
//  `LocationShareEngine` (whose `version` ticks on every decrypted fix), animates the marker
//  between fixes instead of teleporting it, and follows with the camera until you pan.
//
//  Every sharer in the conversation is drawn on THIS one map: a group where three people are
//  sharing is one map with three faces, not three separate frozen bubbles.
//
//  Nothing here changes cadence, encryption or the background mode — this is presentation
//  over the existing decrypted stream.
//
//  iOS 17+ Map(position:) + Annotation. Deployment target is 18.0, so no availability
//  fallback is required. The map is calm and light: POIs excluded, standard style.
//

import SwiftUI
import MapKit
import Combine

struct LocationDetailView: View {
    /// The coordinate the MESSAGE carried. Used for a static pin, and as the fallback for a
    /// live share whose stream has not produced a fix yet.
    let coordinate: CLLocationCoordinate2D
    var label: String?
    var live: Bool = false
    var state: ShareState = .live
    var shareId: String?
    /// Set for a live share so the map can show every sharer in the same chat (G5).
    var conversationId: String?
    /// Accuracy carried by the message itself, for a static pin with no live fix stream.
    var accuracy: Double?

    @Environment(\.dismiss) private var dismiss
    @ObservedObject private var engine = LocationShareEngine.shared
    @ObservedObject private var directory = UserDirectory.shared
    @State private var position: MapCameraPosition
    /// The first manual pan hands the camera to the user; the recenter button takes it back.
    @State private var userPanned = false
    /// Ticks the countdown and the "updated Xs ago" line with no network (the timer guarantee).
    @State private var now = Date()

    private let ticker = Timer.publish(every: 1, on: .main, in: .common).autoconnect()

    init(coordinate: CLLocationCoordinate2D, label: String?, live: Bool,
         state: ShareState, shareId: String?, conversationId: String? = nil,
         accuracy: Double? = nil) {
        self.coordinate = coordinate
        self.label = label
        self.live = live
        self.state = state
        self.shareId = shareId
        self.conversationId = conversationId
        self.accuracy = accuracy
        _position = State(initialValue: .region(MKCoordinateRegion(
            center: coordinate, latitudinalMeters: 800, longitudinalMeters: 800)))
    }

    /// One drawable sharer: who they are, where they are now, and how fresh that is.
    private struct Sharer: Identifiable {
        let id: String            // shareId
        let userId: String
        let coordinate: CLLocationCoordinate2D
        let state: ShareState
        let fixedAt: Date?
        let expiresAt: Date?
        /// Device-reported accuracy of this fix, in metres — drives the honesty line.
        let accuracy: Double?
    }

    /// Every still-running sharer in this conversation, newest fix each. Recomputed whenever
    /// the engine's `version` ticks, which is what makes the pins move.
    private var sharers: [Sharer] {
        guard live else { return [] }
        guard let conversationId else { return primarySharer.map { [$0] } ?? [] }
        let rows = LocationStore.activeInbound(conversationId: conversationId)
        let all: [Sharer] = rows.compactMap { row -> Sharer? in
            guard let fix = engine.lastFix(shareId: row.shareId) else { return nil }
            let s = engine.shareState(shareId: row.shareId, expiresAt: row.expiresAt, cadence: row.cadence)
            guard s != .ended else { return nil }
            return Sharer(id: row.shareId, userId: row.ownerUserId,
                          coordinate: CLLocationCoordinate2D(latitude: fix.lat, longitude: fix.lon),
                          state: s, fixedAt: fix.date, expiresAt: row.expiresAt,
                          accuracy: fix.acc)
        }
        // A share opened from MY OWN bubble is outbound and won't be in `activeInbound`; keep
        // the fallback so the sheet is never empty.
        return all.isEmpty ? (primarySharer.map { [$0] } ?? []) : all
    }

    /// The share this bubble opened — drives the header text and the coordinate readout.
    private var primarySharer: Sharer? {
        guard let shareId else { return nil }
        let fix = engine.lastFix(shareId: shareId)
        return Sharer(
            id: shareId,
            userId: "",
            coordinate: fix.map { CLLocationCoordinate2D(latitude: $0.lat, longitude: $0.lon) } ?? coordinate,
            state: state,
            fixedAt: fix?.date,
            expiresAt: nil,
            accuracy: fix?.acc ?? accuracy)
    }

    /// What the map is centred on: the live fix when we have one, else the message's own.
    private var focus: CLLocationCoordinate2D {
        sharers.first(where: { $0.id == shareId })?.coordinate ?? sharers.first?.coordinate ?? coordinate
    }

    var body: some View {
        ZStack(alignment: .top) {
            map.ignoresSafeArea()
            header
            controls
        }
        // No colour-scheme pin: Peacock tokens resolve per theme, and a sheet that
        // forced light would be the one bright rectangle in a dark app.
        .onReceive(ticker) { now = $0 }
        // Follow the sharer until the user takes the camera over. `focus` changes on every new
        // fix, so this is what makes the camera track a friend who is walking.
        .onChange(of: focus.latitude) { _, _ in follow() }
        .onChange(of: focus.longitude) { _, _ in follow() }
    }

    private func follow() {
        guard live, !userPanned else { return }
        withAnimation(.easeInOut(duration: 0.8)) {
            position = .region(MKCoordinateRegion(center: focus,
                                                  latitudinalMeters: 800, longitudinalMeters: 800))
        }
    }

    private var map: some View {
        Map(position: $position) {
            if sharers.isEmpty {
                Annotation(label ?? (live ? "Live location" : "Location"), coordinate: coordinate) {
                    staticPin
                }
            } else {
                ForEach(sharers) { s in
                    Annotation(s.userId.isEmpty ? (label ?? "Live location")
                                                : directory.displayName(s.userId),
                               coordinate: s.coordinate) {
                        // A face, not a generic dot — the same marker language as the Map tab.
                        // `.animation` on the coordinate makes the pin GLIDE between fixes
                        // (10–15 s apart) instead of teleporting.
                        MapAvatarPin(userId: s.userId.isEmpty ? nil : s.userId,
                                     name: s.userId.isEmpty ? label : directory.displayName(s.userId),
                                     photoURL: s.userId.isEmpty ? nil : directory.photoURL(s.userId),
                                     state: s.state,
                                     size: 44)
                    }
                }
            }
        }
        // Same skin as the Map tab — `.muted` de-emphasises roads and terrain so the avatar
        // pin is the focus. This screen was missing it, so a live share opened into a busier,
        // more saturated map than the tab it was launched from. MapKit's `.standard` follows
        // the environment colorScheme, which ContentView sets from the user's Light/Dark/System
        // choice, so the tiles theme themselves.
        .mapStyle(.standard(elevation: .flat, emphasis: .muted, pointsOfInterest: .excludingAll))
        .animation(.easeInOut(duration: 1.0), value: focus.latitude)
        .animation(.easeInOut(duration: 1.0), value: focus.longitude)
        // Pan detection: a DRAG is unambiguously the user taking the camera over, whereas
        // `onMapCameraChange` also fires for our own programmatic follow moves and would
        // immediately cancel the following it is supposed to protect.
        .simultaneousGesture(DragGesture().onChanged { _ in
            if live && !userPanned { userPanned = true }
        })
    }

    private var staticPin: some View {
        Image(systemName: state == .ended ? "mappin.slash.circle.fill"
                        : (live ? "location.circle.fill" : "mappin.circle.fill"))
            .font(.system(size: 34))
            .foregroundColor(state == .stale ? VoiidColor.textSecondary : VoiidColor.error)
            .shadow(radius: 2)
    }

    private var header: some View {
        HStack(alignment: .top) {
            Button { dismiss() } label: {
                Image(systemName: "xmark")
                    .font(.system(size: 15, weight: .bold)).foregroundColor(VoiidColor.textPrimary)
                    .padding(10).background(VoiidColor.surfaceCard).clipShape(Circle())
                    .shadow(color: .black.opacity(0.15), radius: 3, y: 1)
            }
            Spacer()
            if live {
                VStack(alignment: .trailing, spacing: 3) {
                    HStack(spacing: 6) {
                        Circle().fill(stateColor).frame(width: 8, height: 8)
                        Text(stateLabel).font(VoiidFont.rounded(12, .semibold)).foregroundColor(VoiidColor.textPrimary)
                    }
                    // WhatsApp's "Live until …" — countdown to expiry + freshness of the fix.
                    if let sub = liveSubtitle {
                        Text(sub).font(VoiidFont.rounded(11)).foregroundColor(VoiidColor.textSecondary)
                    }
                }
                .padding(.horizontal, 12).padding(.vertical, 8)
                .background(VoiidColor.surfaceCard).clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
                .shadow(color: .black.opacity(0.12), radius: 3, y: 1)
            }
        }
        .padding(.horizontal, VoiidSpacing.md).padding(.top, VoiidSpacing.md)
    }

    private var controls: some View {
        VStack {
            Spacer()
            if userPanned && live {
                HStack {
                    Spacer()
                    Button {
                        Haptics.tap()
                        userPanned = false
                        follow()
                    } label: {
                        Image(systemName: "location.fill")
                            .font(.system(size: 15, weight: .semibold))
                            .foregroundColor(VoiidColor.primary)
                            .padding(12).background(VoiidColor.surfaceCard).clipShape(Circle())
                            .shadow(color: .black.opacity(0.15), radius: 3, y: 1)
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("Recenter")
                }
                .padding(.horizontal, VoiidSpacing.lg).padding(.bottom, VoiidSpacing.sm)
            }
            // Honesty line (docs/LOCATION.md §10): the marker is an area, not a doorstep, and
            // someone navigating to a friend needs to know the pin carries a radius.
            Text(LocationAccuracy.note(sharers.first(where: { $0.id == shareId })?.accuracy
                                       ?? primarySharer?.accuracy ?? accuracy))
                .font(VoiidFont.rounded(11))
                .foregroundColor(VoiidColor.textPrimary)
                .padding(.horizontal, 12).padding(.vertical, 6)
                .background(VoiidColor.surfaceCard.opacity(0.92))
                .clipShape(Capsule())
                .padding(.bottom, VoiidSpacing.sm)
            HStack(spacing: VoiidSpacing.md) {
                actionButton("Open in Maps", "map.fill") { open(directions: false) }
                actionButton("Directions", "arrow.triangle.turn.up.right.circle.fill") { open(directions: true) }
            }
            .padding(.horizontal, VoiidSpacing.lg).padding(.bottom, VoiidSpacing.xl)
        }
    }

    private func actionButton(_ title: String, _ icon: String, _ tap: @escaping () -> Void) -> some View {
        Button(action: { Haptics.tap(); tap() }) {
            Label(title, systemImage: icon)
                .font(VoiidFont.rounded(15, .semibold)).foregroundColor(VoiidColor.textOnPrimary)
                .frame(maxWidth: .infinity).padding(.vertical, 14)
                .background(VoiidColor.primary).clipShape(Capsule())
        }
        .buttonStyle(.plain)
    }

    /// "ends in 43m · updated 4s ago" — recomputed each tick of `now`.
    private var liveSubtitle: String? {
        guard live else { return nil }
        if state == .ended { return nil }
        var parts: [String] = []
        if let expires = sharers.first(where: { $0.id == shareId })?.expiresAt, expires > now {
            parts.append("ends in \(Self.duration(expires.timeIntervalSince(now)))")
        }
        if state == .stale {
            parts.append("may have lost signal")
        } else if let fixedAt = sharers.first(where: { $0.id == shareId })?.fixedAt
                    ?? primarySharer?.fixedAt {
            parts.append("updated \(Self.age(now.timeIntervalSince(fixedAt)))")
        } else {
            parts.append("waiting for first fix")
        }
        return parts.isEmpty ? nil : parts.joined(separator: " · ")
    }

    private static func duration(_ t: TimeInterval) -> String {
        let mins = max(0, Int(t) / 60)
        return mins >= 60 ? "\(mins / 60)h \(mins % 60)m" : "\(mins)m"
    }

    private static func age(_ t: TimeInterval) -> String {
        let secs = max(0, Int(t))
        switch secs {
        case ..<10: return "just now"
        case ..<60: return "\(secs)s ago"
        case ..<3600: return "\(secs / 60)m ago"
        default: return "\(secs / 3600)h ago"
        }
    }

    private var stateColor: Color {
        switch state { case .live: return VoiidColor.success
                       case .stale: return VoiidColor.warning
                       case .ended: return VoiidColor.textSecondary }
    }
    private var stateLabel: String {
        switch state { case .live: return "Live"; case .stale: return "May have lost signal"; case .ended: return "Ended" }
    }

    /// System handoff — no in-app routing. Opens Apple Maps at the coordinate.
    private func open(directions: Bool) {
        let target = focus
        let placemark = MKPlacemark(coordinate: target)
        let item = MKMapItem(placemark: placemark)
        item.name = label ?? (live ? "Live location" : "Shared location")
        item.openInMaps(launchOptions: directions
            ? [MKLaunchOptionsDirectionsModeKey: MKLaunchOptionsDirectionsModeDriving] : nil)
    }
}

File: apps/android/app/src/main/java/com/voiid/app/main/LocationDetailView.kt
package com.voiid.app.main

import androidx.activity.compose.BackHandler
import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Directions
import androidx.compose.material.icons.filled.Map
import androidx.compose.material.icons.filled.MyLocation
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import com.google.maps.android.compose.CameraMoveStartedReason
import com.google.maps.android.compose.GoogleMap
import com.google.maps.android.compose.MapType
import com.google.maps.android.compose.MapUiSettings
import com.google.maps.android.compose.MarkerComposable
import com.google.maps.android.compose.MarkerState
import com.google.maps.android.compose.rememberCameraPositionState
import com.voiid.app.BuildConfig
import com.voiid.app.model.LocationEnvelope
import com.voiid.app.model.LiveShareView
import com.voiid.app.model.ShareState
import com.voiid.app.net.ChatEngine
import com.voiid.app.net.LocationShareEngine
import com.voiid.app.store.UserDirectory
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import com.voiid.app.ui.theme.VoiidColor
import com.voiid.app.ui.theme.VoiidFont
import com.voiid.app.ui.theme.VoiidRadius

/**
 * Full-screen location detail (docs/LOCATION.md §4): the map, Open in Maps and Directions
 * handoffs. No in-app routing — we hand off to the system map app (§10.10).
 *
 * LIVE, not a snapshot. This view used to render `ref.lat/ref.lon` once, so a friend walking
 * across town stayed pinned wherever they were when you opened the sheet — the pipeline was
 * delivering fixes the whole time and only the bubble showed them. It now observes
 * [LocationShareEngine.inboundViews] (a snapshot state map, so new fixes recompose us), moves
 * the marker with a short animation rather than teleporting it, and follows with the camera
 * until you pan.
 *
 * Every sharer in the conversation is drawn on THIS one map (a group where three people are
 * sharing is one map with three faces, not three separate frozen bubbles).
 *
 * Nothing here changes cadence, encryption or the foreground service — this is presentation
 * over the existing decrypted stream.
 */
@Composable
fun LocationDetailView(ref: ChatEngine.LocationRef, conversationId: String? = null, onClose: () -> Unit) {
    val context = LocalContext.current
    BackHandler { onClose() }

    val isLive = ref.kind == LocationEnvelope.K_LIVE_START

    // 1s ticker: drives the countdown, the "updated Xs ago" line and the LIVE→STALE→ENDED
    // transition WITHOUT any network — the timer guarantee (§3) holds offline.
    var now by remember { mutableLongStateOf(System.currentTimeMillis()) }
    LaunchedEffect(isLive) {
        if (!isLive) return@LaunchedEffect
        while (true) { kotlinx.coroutines.delay(1_000); now = System.currentTimeMillis() }
    }

    // Every live sharer in this conversation (G5). Falls back to just this share when we were
    // opened without a conversation id, and to an empty list for a static pin.
    val shares: List<LiveShareView> = when {
        !isLive -> emptyList()
        conversationId != null -> LocationShareEngine.inboundForConversation(conversationId, now)
        else -> listOfNotNull(LocationShareEngine.inbound(ref.shareId))
    }
    // The share this bubble opened, kept first so the header describes the right person.
    val primary = shares.firstOrNull { it.shareId == ref.shareId } ?: shares.firstOrNull()

    // Most current coordinate: the live fix if the stream has produced one, else the
    // coordinate the message itself carried.
    val lat = primary?.lastFix?.lat ?: ref.lat
    val lon = primary?.lastFix?.lon ?: ref.lon
    if (lat == null || lon == null) { onClose(); return }

    val state = primary?.state(now) ?: if (isLive) ShareState.ENDED else null

    Column(Modifier.fillMaxSize().background(VoiidColor.background)) {
        Row(
            Modifier.fillMaxWidth().statusBarsPadding().padding(horizontal = 16.dp, vertical = 12.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Icon(Icons.Default.Close, "Close", tint = VoiidColor.textPrimary, modifier = Modifier.size(26.dp).clickable { onClose() })
            Column(Modifier.weight(1f)) {
                Text(
                    ref.label?.ifBlank { null } ?: if (isLive) "Live location" else "Location",
                    style = VoiidFont.rounded(18, FontWeight.SemiBold), color = VoiidColor.textPrimary,
                )
                // WhatsApp's "Live until …": expiry countdown + freshness of the last fix.
                if (isLive) {
                    Text(
                        liveSubtitle(state, primary, ref.expiresAt, now),
                        style = VoiidFont.rounded(12), color = subtitleColor(state), maxLines = 1,
                    )
                }
            }
            if (isLive && state != null) LiveDot(state)
        }

        Box(Modifier.fillMaxWidth().weight(1f)) {
            if (!BuildConfig.MAPS_CONFIGURED) {
                MapUnavailableCard(Modifier.fillMaxSize(), lat, lon)
            } else {
                LiveDetailMap(shares = shares, fallback = LatLng(lat, lon), state = state, now = now)
            }
        }

        Column(
            Modifier.fillMaxWidth().background(VoiidColor.surfaceCard).navigationBarsPadding().padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            Text("%.5f, %.5f".format(lat, lon), style = VoiidFont.rounded(13, FontWeight.Medium), color = VoiidColor.textSecondary)
            // Honesty line (docs/LOCATION.md §10): a phone GPS fix is not a pinpoint, and a
            // reader who treats it as one can walk to the wrong door. Prefer the accuracy the
            // DEVICE actually reported for this fix; fall back to a conservative typical value
            // when a sender's payload carried none.
            Text(
                accuracyNote(primary?.lastFix?.acc ?: ref.acc),
                style = VoiidFont.rounded(11), color = VoiidColor.textSecondary,
            )
            Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                DetailAction(Icons.Default.Map, "Open in Maps", Modifier.weight(1f)) { openInMaps(context, lat, lon, ref.label) }
                DetailAction(Icons.Default.Directions, "Directions", Modifier.weight(1f)) { openDirections(context, lat, lon, ref.label) }
            }
        }
    }
}

/**
 * The interactive map: one animated avatar marker per sharer, camera following until the user
 * takes over.
 *
 * FOLLOW RULE: we recentre on every new fix only while [userPanned] is false. The first
 * gesture-driven camera move sets it (so the map stops fighting the user's finger) and the
 * recenter FAB clears it — the WhatsApp behaviour.
 */
@Composable
private fun LiveDetailMap(
    shares: List<LiveShareView>,
    fallback: LatLng,
    state: ShareState?,
    now: Long,
) {
    val camera = rememberCameraPositionState {
        position = CameraPosition.fromLatLngZoom(fallback, 16f)
    }
    var userPanned by remember { mutableStateOf(false) }
    // Only a GESTURE counts as taking over; our own animateCamera calls must not disable follow.
    LaunchedEffect(camera.cameraMoveStartedReason) {
        if (camera.cameraMoveStartedReason == CameraMoveStartedReason.GESTURE) userPanned = true
    }

    // Positions we actually draw, one per sharer (falls back to the message coordinate when the
    // stream hasn't produced a fix yet).
    val points: List<Pair<LiveShareView, LatLng>> = shares.mapNotNull { s ->
        s.lastFix?.let { s to LatLng(it.lat, it.lon) }
    }
    val targets = points.map { it.second }.ifEmpty { listOf(fallback) }

    // Follow: a single sharer centres; several fit all of them in view.
    LaunchedEffect(targets, userPanned) {
        if (userPanned) return@LaunchedEffect
        if (targets.size == 1) {
            camera.animate(CameraUpdateFactory.newLatLng(targets.first()), 800)
        } else {
            val bounds = LatLngBounds.builder().apply { targets.forEach { include(it) } }.build()
            runCatching { camera.animate(CameraUpdateFactory.newLatLngBounds(bounds, 160), 800) }
        }
    }

    Box(Modifier.fillMaxSize()) {
        GoogleMap(
            modifier = Modifier.fillMaxSize(),
            cameraPositionState = camera,
            uiSettings = MapUiSettings(zoomControlsEnabled = false, compassEnabled = false, mapToolbarEnabled = false),
            properties = com.google.maps.android.compose.MapProperties(mapType = MapType.NORMAL),
        ) {
            if (points.isEmpty()) {
                MarkerComposable(state = MarkerState(position = fallback)) {
                    AvatarPin(userId = null, stale = state == ShareState.STALE, ended = state == ShareState.ENDED)
                }
            } else {
                for ((share, target) in points) {
                    val s = share.state(now)
                    AnimatedAvatarMarker(
                        target = target,
                        userId = share.ownerUserId,
                        stale = s == ShareState.STALE,
                        ended = s == ShareState.ENDED,
                    )
                }
            }
        }

        // Recenter: only offered once the user has taken the camera over.
        if (userPanned) {
            Box(
                Modifier
                    .align(Alignment.BottomEnd)
                    .padding(16.dp)
                    .size(44.dp)
                    .clip(CircleShape)
                    .background(VoiidColor.surfaceCard)
                    .clickable { userPanned = false },
                contentAlignment = Alignment.Center,
            ) {
                Icon(Icons.Default.MyLocation, "Recenter", tint = VoiidColor.primary, modifier = Modifier.size(22.dp))
            }
        }
    }
}

/**
 * A marker that GLIDES to each new fix instead of teleporting.
 *
 * Fixes arrive every 10–15 s, so an un-animated marker reads as a series of jumps. We
 * interpolate lat/lon over ~1 s — comfortably shorter than the cadence, so a marker always
 * settles before the next fix and never lags behind reality.
 */
@Composable
private fun AnimatedAvatarMarker(target: LatLng, userId: String?, stale: Boolean, ended: Boolean) {
    val latAnim = remember { Animatable(target.latitude.toFloat()) }
    val lonAnim = remember { Animatable(target.longitude.toFloat()) }
    LaunchedEffect(target) {
        val spec = tween<Float>(durationMillis = 1_000, easing = LinearEasing)
        coroutineScope {
            launch { latAnim.animateTo(target.latitude.toFloat(), spec) }
            launch { lonAnim.animateTo(target.longitude.toFloat(), spec) }
        }
    }
    val pos = LatLng(latAnim.value.toDouble(), lonAnim.value.toDouble())
    MarkerComposable(
        keys = arrayOf(userId ?: "", stale, ended),
        state = MarkerState(position = pos),
        title = userId?.let { UserDirectory.displayName(it) },
    ) {
        AvatarPin(userId = userId, stale = stale, ended = ended)
    }
}

@Composable
private fun LiveDot(state: ShareState) {
    Box(
        Modifier.size(10.dp).clip(CircleShape).background(
            when (state) {
                ShareState.LIVE -> VoiidColor.primary
                ShareState.STALE -> VoiidColor.textSecondary
                ShareState.ENDED -> VoiidColor.textSecondary.copy(alpha = 0.5f)
            },
        ),
    )
}

/** "Live · ends in 43 min · updated 4s ago" — the freshness WhatsApp shows on its detail map. */
private fun liveSubtitle(state: ShareState?, share: LiveShareView?, expiresAt: Long?, now: Long): String {
    if (state == ShareState.ENDED) return "Live location ended"
    val parts = mutableListOf<String>()
    val ends = share?.expiresAt ?: expiresAt
    if (ends != null && ends > now) parts.add("ends in ${humanDuration(ends - now)}")
    val fixedAt = share?.lastFix?.fixedAt
    parts.add(
        when {
            state == ShareState.STALE -> "may have lost signal"
            fixedAt != null -> "updated ${humanAge(now - fixedAt)}"
            else -> "waiting for first fix"
        },
    )
    return parts.joinToString(" · ")
}

private fun subtitleColor(state: ShareState?) =
    if (state == ShareState.STALE || state == ShareState.ENDED) VoiidColor.textSecondary else VoiidColor.primary

private fun humanDuration(ms: Long): String {
    val mins = (ms / 60_000L).coerceAtLeast(0)
    return if (mins >= 60) "${mins / 60}h ${mins % 60}m" else "${mins}m"
}

private fun humanAge(ms: Long): String {
    val secs = (ms / 1000L).coerceAtLeast(0)
    return when {
        secs < 10 -> "just now"
        secs < 60 -> "${secs}s ago"
        secs < 3600 -> "${secs / 60}m ago"
        else -> "${secs / 3600}h ago"
    }
}

/**
 * "Accurate to about N m" — the one place this sentence is produced, so the bubble, the
 * full-screen detail and the map card cannot drift apart.
 *
 * WHY IT EXISTS: a marker drawn as a single point reads as an exact doorstep, and it is not.
 * A phone fix is typically 10–30 m in the open and much worse indoors or between tall
 * buildings, so a viewer navigating to a friend needs to know the pin has a radius. We show
 * the accuracy the SENDER'S DEVICE reported for that fix rather than a fixed number, because
 * inventing a figure would be its own dishonesty; [FALLBACK_ACCURACY_M] covers only the case
 * where a payload carried no accuracy at all.
 *
 * Note this is separate from the coordinate rounding we apply (5 dp ≈ 1 m) — that is a privacy
 * measure and is far finer than the fix's own error, so it never dominates this number.
 */
private const val FALLBACK_ACCURACY_M = 30

fun accuracyNote(acc: Double?): String {
    val metres = acc?.takeIf { it.isFinite() && it > 0 }?.toInt() ?: FALLBACK_ACCURACY_M
    // Round to something a human reads as an estimate, never a false precision like "37 m".
    val rounded = when {
        metres <= 10 -> 10
        metres <= 30 -> ((metres + 4) / 5) * 5
        metres <= 100 -> ((metres + 9) / 10) * 10
        else -> ((metres + 49) / 50) * 50
    }
    return "Accurate to about $rounded m — GPS is approximate"
}

/** Directions handoff — shared with the map tab's place card, so the intent lives in one place. */
fun openDirections(context: android.content.Context, lat: Double, lon: Double, label: String?) {
    runCatching {
        context.startActivity(
            android.content.Intent(android.content.Intent.ACTION_VIEW, android.net.Uri.parse("google.navigation:q=$lat,$lon"))
                .addFlags(android.content.Intent.FLAG_ACTIVITY_NEW_TASK),
        )
    }.onFailure { openInMaps(context, lat, lon, label) }
}

@Composable
private fun DetailAction(icon: androidx.compose.ui.graphics.vector.ImageVector, label: String, modifier: Modifier, onClick: () -> Unit) {
    Row(
        modifier
            .clip(RoundedCornerShape(VoiidRadius.pill))
            .background(VoiidColor.fieldFill)
            .clickable { onClick() }
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Icon(icon, null, tint = VoiidColor.primary, modifier = Modifier.size(18.dp))
        Text(label, style = VoiidFont.rounded(14, FontWeight.SemiBold), color = VoiidColor.textPrimary)
        Spacer(Modifier.height(0.dp))
    }
}
